### PR TITLE
[feature](nereids)support mark join

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes

support mark join
be part is implement by Author: Jerry Hu <[mrhhsg@gmail.com](mailto:mrhhsg@gmail.com)>

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 1]
```diff
@@ -322,7 +338,8 @@ private ImmutableList<RelatedInfo> collectRelatedInfo(List<Expression> subqueryC
     private LogicalPlan subqueryToApply(List<SubqueryExpr> subqueryExprs, LogicalPlan childPlan,
                                         Map<SubqueryExpr, Optional<MarkJoinSlotReference>> subqueryToMarkJoinSlot,
                                         CascadesContext ctx,
-                                        Optional<Expression> conjunct, boolean isProject) {
+                                        Optional<Expression> conjunct, boolean isProject,
+                                        boolean isMarkJoinSlotNotNull) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:342]: add more comment to explain the effect of isMarkJoinSlotNotNull

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 2]
```diff
@@ -1270,7 +1292,7 @@ public PlanFragment visitPhysicalHashJoin(
         List<SlotDescriptor> rightIntermediateSlotDescriptor = Lists.newArrayList();
         TupleDescriptor intermediateDescriptor = context.generateTupleDesc();
 
-        if (hashJoin.getOtherJoinConjuncts().isEmpty()
+        if (hashJoin.getOtherJoinConjuncts().isEmpty() && (isHashJoinConjunctsEmpty ^ isMarkJoinConjunctsEmpty)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:1295]:
```suggestion
        if (hashJoin.getOtherJoinConjuncts().isEmpty() && (isHashJoinConjunctsEmpty != isMarkJoinConjunctsEmpty)
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 3]
```diff
@@ -216,12 +226,18 @@ public List<Rule> buildRules() {
...
                         SubqueryContext context = new SubqueryContext(subqueryExprs);
                         Expression conjunct = replaceSubquery.replace(subqueryConjuncts.get(i), context);
-
+                        boolean isMarkSlotNotNull = conjunct.containsType(MarkJoinSlotReference.class)
+                                        ? ExpressionUtils
+                                                .canInferNotNullForMarkSlot(
+                                                        TrySimplifyPredicateWithMarkJoinSlot.INSTANCE
+                                                                .rewrite(conjunct,
+                                                                        null))
+                                        : false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:235]: nit: optimize the wrap

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalApply.java [hunk 4]
```diff
@@ -62,13 +62,16 @@ public class LogicalApply<LEFT_CHILD_TYPE extends Plan, RIGHT_CHILD_TYPE extends
     // Whether adding the subquery's output to projects
     private final boolean needAddSubOutputToProjects;
 
+    private final boolean isMarkJoinSlotNotNull;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalApply.java:65]: explain it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/UsingJoin.java [hunk 5]
```diff
@@ -47,6 +46,7 @@ public class UsingJoin<LEFT_CHILD_TYPE extends Plan, RIGHT_CHILD_TYPE extends Pl
     private final JoinType joinType;
     private final ImmutableList<Expression> otherJoinConjuncts;
     private final ImmutableList<Expression> hashJoinConjuncts;
+    private final ImmutableList<Expression> markJoinConjuncts;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/UsingJoin.java:49]: why update UsingJoin? it could not be mark join at all

**Comment by starocean999** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by keanji-x — APPROVED** [review]

**Pull request merged**